packages/sanity/src/core/store/_legacy/history/useTimelineStore.ts

```diff
@@ -0,0 +1,209 @@
+}
+
+/** @internal */
+export interface TimelineStore {
+  findRangeForRev: TimelineController['findRangeForNewRev']
+  findRangeForSince: TimelineController['findRangeForNewSince']
+  loadMore: () => void
+  getSnapshot: () => TimelineState
+  subscribe: (callback: () => void) => () => void
+}
+
+/**
+ * Creates a store which handles the creation of a document Timeline,
+ * TimelineController and also fetches pre-requisite document snapshots.
+ *
+ * `TimelineStore` exposes select TimelineController methods used to query
+ * ranges and fetch more transactions. It can also be used with
+ * `useSyncExternalStore` to subscribe to selected state changes.
+ *
+ * @internal
+ * */
+export function useTimelineStore({
+  documentId,
+  documentType,
+  onError,
+  rev,
+  since,
+}: UseTimelineControllerOpts): TimelineStore {
+  const historyStore = useHistoryStore()
+  const snapshotsSubscriptionRef = useRef<Subscription | null>(null)
+  const timelineStateRef = useRef<TimelineState>(INITIAL_TIMELINE_STATE)
+  const client = useClient(DEFAULT_STUDIO_CLIENT_OPTIONS)
+
+  /**
+   * The mutable TimelineController, used internally
+   */
+  const controller = useMemo(
+    () =>
+      historyStore.getTimelineController({
+        client,
+        documentId,
+        documentType,
+      }),
+    [client, documentId, documentType, historyStore]
+  )
+
+  /**
+   * A BehaviorSubject which can be subscribed by multiple observers, which broadcasts
+   * the the latest state of the (mutable) TimelineController.
+   */
+  const timelineController$ = useMemo(
+    () => new BehaviorSubject<TimelineController>(controller),
+    [controller]
+  )
+
+  /**
+   * Broadcasts the updated state on TimelineController's `handler` updates (as well as on mount / unmount).
+   *
+   * Note that TimelineController triggers updates very frequently when fetching transactions, which it does so
+   * in fairly small batches.
+   *
+   * This will also suspend TimelineController transaction fetching on cleanup, ensuring that fetches don't
+   * continue once the document pane has been closed.
+   */
+  useEffect(() => {
+    controller.setRange(since || null, rev || null)
+    timelineController$.next(controller)
+
+    controller.handler = (err, innerController) => {
+      if (err) {
+        timelineController$.error(err)
+      } else {
+        /**
+         * NOTE: TimelineController requires that you call `setRange` manually whenever its internal
+         * timeline has changed (e.g. has a result of fetched transactions).
+         * Ideally, TimelineController would be updated to handle this automatically for us.
+         * Until then, this workaround waits until the next call stack before calling `setRange`.
+         */
+        setTimeout(() => {
+          innerController.setRange(since || null, rev || null)
+          timelineController$.next(innerController)
+        }, 0)
+      }
+    }
+    controller.resume()
+    return () => controller.suspend()
```

Comment:
Will this potentially re-use controllers when navigating between workspaces, or will it be suspended forever and a new one created if you come back to a previous workspace?

Reply:
It won't – controllers are created on a per-document basis and are never reused. Switching workspaces will unmount the document pane, suspend any future transaction fetching and create a new controller instance.

This was prior behaviour too, though can certainly see a future where this instance is re-used and shared throughout.